File: examples/python-api/app/application/user/use_cases.py
"""User Use Cases."""

from app.application.user.dtos import CreateUserDto, UserResponseDto
from app.domain.user.entity import User
from app.domain.user.repository import UserRepository


class CreateUserUseCase:
    def __init__(self, user_repository: UserRepository):
        self.user_repository = user_repository

    async def execute(self, dto: CreateUserDto) -> UserResponseDto:
        normalized_email = dto.email.strip().lower()
        normalized_name = dto.name.strip()
        if not normalized_name:
            raise ValueError("Name is required")

        # 1. 비즈니스 검증: 이메일 중복 확인
        existing = await self.user_repository.find_by_email(normalized_email)
        if existing:
            raise ValueError("Email already exists")

        # 2. 도메인 엔터티 생성
        user = User.create(email=normalized_email, name=normalized_name)

        # 3. 영속화
        await self.user_repository.save(user)

        # 4. 응답 DTO 변환
        return UserResponseDto.from_entity(user)


class GetUserUseCase:
    def __init__(self, user_repository: UserRepository):
        self.user_repository = user_repository

    async def execute(self, user_id: str) -> UserResponseDto:
        normalized_id = user_id.strip()
        if not normalized_id:
            raise ValueError("User ID is required")
        user = await self.user_repository.find_by_id(normalized_id)
        if not user:
            raise ValueError(f"User not found: {normalized_id}")
        return UserResponseDto.from_entity(user)

File: examples/typescript-api/src/application/user/use-cases.ts
import { User } from '../../domain/user/entity';
import { IUserRepository } from '../../domain/user/repository';
import {
  UserNotFoundError,
  UserAlreadyExistsError,
} from '../../domain/user/errors';
import {
  CreateUserInput,
  UpdateUserInput,
  UserResponse,
  ListUsersQuery,
  ListUsersResponse,
} from './dto';
import { randomUUID } from 'crypto';

/**
 * User Use Cases
 *
 * 애플리케이션 비즈니스 로직을 담당합니다.
 */
export class UserUseCases {
  constructor(private readonly userRepository: IUserRepository) {}

  /**
   * 사용자 생성
   */
  async createUser(input: CreateUserInput): Promise<UserResponse> {
    // 이메일 중복 확인
    const existing = await this.userRepository.findByEmail(input.email);
    if (existing) {
      throw new UserAlreadyExistsError(input.email);
    }

    // 사용자 생성
    const user = User.create({
      id: randomUUID(),
      email: input.email,
      name: input.name,
    });

    // 저장
    const saved = await this.userRepository.save(user);

    return this.toResponse(saved);
  }

  /**
   * 사용자 조회
   */
  async getUser(id: string): Promise<UserResponse> {
    const user = await this.userRepository.findById(id);
    if (!user) {
      throw new UserNotFoundError(id);
    }

    return this.toResponse(user);
  }

  /**
   * 사용자 목록 조회
   */
  async listUsers(query: ListUsersQuery): Promise<ListUsersResponse> {
    const [users, total] = await Promise.all([
      this.userRepository.findAll({
        limit: query.limit,
        offset: query.offset,
        status: query.status,
      }),
      this.userRepository.count({ status: query.status }),
    ]);

    return {
      data: users.map((user) => this.toResponse(user)),
      meta: {
        total,
        limit: query.limit,
        offset: query.offset,
      },
    };
  }

  /**
   * 사용자 업데이트
   */
  async updateUser(id: string, input: UpdateUserInput): Promise<UserResponse> {
    const user = await this.userRepository.findById(id);
    if (!user) {
      throw new UserNotFoundError(id);
    }

    // 이메일 변경 시 중복 확인
    if (input.email && input.email !== user.email) {
      const existing = await this.userRepository.findByEmail(input.email);
      if (existing) {
        throw new UserAlreadyExistsError(input.email);
      }
    }

    const updated = user.update(input);
    const saved = await this.userRepository.save(updated);

    return this.toResponse(saved);
  }

  /**
   * 사용자 삭제
   */
  async deleteUser(id: string): Promise<void> {
    const user = await this.userRepository.findById(id);
    if (!user) {
      throw new UserNotFoundError(id);
    }

    await this.userRepository.delete(id);
  }

  /**
   * Entity to Response 변환
   */
  private toResponse(user: User): UserResponse {
    return {
      id: user.id,
      email: user.email,
      name: user.name,
      status: user.status,
      createdAt: user.createdAt.toISOString(),
      updatedAt: user.updatedAt.toISOString(),
    };
  }
}
